python/src/iceberg/conversions.py

```diff
@@ -0,0 +1,294 @@
+    are defined here as generic functions using the @singledispatch decorator. For each PrimitiveType
+    implementation, a concrete function is registered for each generic conversion function. For PrimitiveType
+    implementations that share the same conversion logic, registrations can be stacked.
+"""
+import struct
+import uuid
+from decimal import Decimal, localcontext
+from functools import singledispatch
+from typing import Union
+
+from iceberg.types import (
+    BinaryType,
+    BooleanType,
+    DateType,
+    DecimalType,
+    DoubleType,
+    FixedType,
+    FloatType,
+    IntegerType,
+    LongType,
+    PrimitiveType,
+    StringType,
+    TimestampType,
+    TimestamptzType,
+    TimeType,
+    UUIDType,
+)
+
+
+def convert_decimal_to_unscaled_value(decimal_type_var: DecimalType, value: Decimal) -> int:
+    """Get an unscaled value given a Decimal value and a DecimalType instance with defined precision and scale
+
+    Args:
+        decimal_type_var (DecimalType): A DecimalType instance with precision and scale
+        value (Decimal): A Decimal instance
+
+    Raises:
+        ValueError: If either the scale or precision of `value` does not match that defined in the DecimalType instance provided as `decimal_type_var`
+
+    Returns:
+        int: The unscaled value
+    """
+    value_as_tuple = value.as_tuple()
+    value_precision = len(value_as_tuple.digits)
+    value_scale = -value_as_tuple.exponent
+
+    if value_scale != decimal_type_var.scale:
+        raise ValueError(f"Cannot serialize value, scale of value does not match type {decimal_type_var}: {value_scale}")
+    elif value_precision > decimal_type_var.precision:
+        raise ValueError(
+            f"Cannot serialize value, precision of value is greater than precision of type {decimal_type_var}: {value_precision}"
+        )
+
+    with localcontext() as ctx:
+        ctx.prec = decimal_type_var.precision
+        sign, digits, exponent = value.as_tuple()
+        value_w_adjusted_scale = Decimal((sign, digits, exponent + decimal_type_var.scale))
+        return int(value_w_adjusted_scale.to_integral_value())
+
+
+@singledispatch
+def from_partition_value_to_py(primitive_type, partition_value_as_str: str):
+    """A generic function function which converts a partition string to a python built-in
+
+    Args:
+        primitive_type(PrimitiveType): An implementation of the PrimitiveType base class
+        partition_value_as_str(str): A string representation of a partition value
+    """
+    if partition_value_as_str is None:
+        return None
+    raise TypeError(
+        f"Cannot convert partition string to python built-in, type {primitive_type} not supported: '{partition_value_as_str}'"
+    )
+
+
+@from_partition_value_to_py.register(BooleanType)
+def _(primitive_type, partition_value_as_str: str):
+    if partition_value_as_str is None:
+        return None
```

Comment:
Instead of adding this check in every method, could we use an outer function that calls the `@singledispatch` method?

```python
def partition_to_py(primitive_type: PrimitiveType, value_str: str):
    if value_str is None:
        return None
    elif value_str == '__HIVE_DEFAULT_PARTITION__':
        return None
   return _partition_to_py(primitive_type, value_str)

@singledispatch
def _partition_to_py(primitive_type: PrimitiveType, value_str: str):
    raise TypeError("Cannot convert '{value_str}' to unsupported type: {primitive_type}")
```

Reply:
Another option is to create a simple `@handle_none` decorator and decorate the relevant functions. I think this may be a bit better when navigating the code in the future. When using a "Go to Definition" somehwere `partition_to_py` is used, it will take you straight to the definition without the double hop and I think most people would ignore the `@handle_none` decorator unless it sounds relevant to what they're looking for. (see here and let me know what you think)
```py
def handle_none(func):
    """A decorator function to handle cases where partition values are `None` or "__HIVE_DEFAULT_PARTITION__"

    Args:
        func (Callable): A function registered to the singledispatch function `partition_to_py`
    """

    def wrapper(primitive_type, partition_value_as_str):
        if partition_value_as_str is None:
            return None
        elif partition_value_as_str == "__HIVE_DEFAULT_PARTITION__":
            return None
        return func(primitive_type, partition_value_as_str)

    return wrapper


@partition_to_py.register(BooleanType)
@handle_none
def _(primitive_type, value_str: str) -> Union[int, float, str, uuid.UUID]:
    return value_str.lower() == "true"
```